add `torch.testing` to docs

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56373 add `torch.testing` to docs**
* #56365 Split `test_testing.py::TestAsserts` for multiple devices

To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

pmeier committed:
add `torch.testing` to docs

To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

[ghstack-poisoned]

pmeier commented on docs/source/index.rst:
```diff
@@ -81,6 +81,7 @@ Features described in this documentation are classified by release status:
    torch.random <random>
    sparse
    storage
+   torch.testing <testing>
```
I'm really not sure where to put this here. It seems kinda alphabetical, so IMO that would be the right place. Feel free to suggest a better position.

mruberry replied:
This seems good

mattip commented:
This seems to need some kind of a rebase. Building the [pytorch_linux_xenial_py3_6_gcc5_4_build](https://app.circleci.com/pipelines/github/pytorch/pytorch/304523/workflows/607f6f83-c83b-4e8d-8334-6e305c65862a/jobs/12557013) failed with
```
Auto-merging test/test_testing.py
CONFLICT (content): Merge conflict in test/test_testing.py
Automatic merge failed; fix conflicts and then commit the result.

```

Since that is a [prerequisite to the doc build](https://app.circleci.com/pipelines/github/pytorch/pytorch/304523/workflows/607f6f83-c83b-4e8d-8334-6e305c65862a), the docs cannot be previewed.

pmeier committed:
Update on "add `torch.testing` to docs"


To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

[ghstack-poisoned]

pmeier commented:
> This seems to need some kind of a rebase.

I've just rebased on `master` and there were no conflicts. Lets hope CI is happy now.

pmeier commented on docs/source/testing.rst:
```diff
@@ -0,0 +1,7 @@
+torch.testing
+=============
+
```
Do we need a disclaimer about the beta status here?

mruberry replied:
Yes!

We should probably steal something like what torch.linalg has:

"Warning! This module is in BETA. New functions are still being added, and existing functions may change in future PyTorch releases."

mruberry reviewed (approved):
Cool! This just needs the BETA warning.

After this stack, however, we should carefully review the public-facing docs holistically.

pmeier committed:
Update on "add `torch.testing` to docs"


To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

[ghstack-poisoned]

pmeier committed:
Update on "add `torch.testing` to docs"


To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

[ghstack-poisoned]

pmeier commented on docs/source/testing.rst:
```diff
@@ -0,0 +1,12 @@
+
+.. warning::
+
+    This module is in BETA. New functions are still being added, and some functions may change in future PyTorch
+    releases. See the documentation of each function for details.
```
@mruberry Since we opted for releasing in a prototype state instead, any recommendations on the wording?

```suggestion
    This module is in a PROTOTYPE state. New functions are still being added, and the available functions may change in future PyTorch releases. We are actively looking for feedback for UI/UX improvements or missing functionalities.
```

mruberry replied:
Let's just wait on this part of the stack until we get more feedback.

pmeier committed:
Update on "add `torch.testing` to docs"


To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

[ghstack-poisoned]

pmeier committed:
Update on "add `torch.testing` to docs"


To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

[ghstack-poisoned]

pmeier committed:
Update on "add `torch.testing` to docs"


To reviewers: **please be nitpicky**. I've read this so often that I probably missed some typos and inconsistencies.

[ghstack-poisoned]

pmeier commented:
Superseded by #57247.
